File: samples/leetcode/arrays/valid_anagram.cpp
#include <string>
#include <iostream>
#include <unordered_map>
#include <cassert>

using namespace std;

class Solution {
public:
    bool isAnagram(string s, string t) {
        auto s_map = unordered_map<char, int>{};
        auto t_map = unordered_map<char, int>{};
        for (const auto &c : s) {
            if (s_map.contains(c)) {
                ++s_map[c];
            } else {
                s_map[c] = 0;
            }
        }
        for (const auto &c : t) {
            if (t_map.contains(c)) {
                ++t_map[c];
            } else {
                t_map[c] = 0;
            }
        }
        for (auto [key,value] : s_map) {
            if (!t_map.contains(key)) return false;
            if (s_map.at(key) != t_map.at(key)) return false;
        }
        return true;
    }
};

int main() {
    std::string s("anagram");
    std::string t("nagaram");

    auto solution = Solution();

    assert(solution.isAnagram(s, t));

    s = "rat";
    t = "car";

    assert(!solution.isAnagram(s, t));
    
    return 0;
}

File: samples/leetcode/arrays/valid_anagram.rs
use std::collections::HashMap;
use std::error::Error;

struct Solution;

impl Solution {
    pub fn is_anagram(s: String, t: String) -> bool {
        let mut s_map: HashMap<char, i32> = HashMap::new();
        let mut t_map: HashMap<char, i32> = HashMap::new();

        for c in s.chars() {
            let count = s_map.entry(c).or_insert(1);
            *count += 1;
        }

        for c in t.chars() {
            let count = t_map.entry(c).or_insert(1);
            *count += 1;
        }

        // Check if s_map and t_map have the same keys with the same values
        if s_map.len() != t_map.len() {
            return false;
        }

        for (key, value) in &s_map {
            if t_map.get(key) != Some(value) {
                return false;
            }
        }

        true
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    let s : String = String::from("anagram");
    let t : String = String::from("nagaram");

    assert!(Solution::is_anagram(s, t), "solution is incorrect");

    Ok(())
}